app/models/piece.rb

```diff
@@ -1,9 +1,62 @@
 class Piece < ApplicationRecord
   belongs_to :game
+
+  PAWN = 'Pawn'.freeze
+  ROOK = 'Rook'.freeze
+  KNIGHT = 'Knight'.freeze
+  BISHOP = 'Bishop'.freeze
+  QUEEN = 'Queen'.freeze
+  KING = 'King'.freeze
+
+  def obstructed?(destination_x, destination_y)
+    return true if self.invalid?(destination_x, destination_y)
+    return true if self.horizontal_or_vertical_obstruction?(destination_x, destination_y)
+    return true if self.diagonal_obstruction?(destination_x, destination_y)
+    false
+  end
+
+  def horizontal_or_vertical?(destination_x, destination_y)
+    return false unless destination_x == self.x_position || destination_y == self.y_position
+  end
+
+  def diagonal?(destination_x, destination_y)
+    # in order for the move to be diagonal, the piece must be moving by the same distance both horizontally and vertically.
+    return false unless (destination_x - self.x_position).abs == (destination_y - self.y_position).abs
+  end
+
+  def invalid?(destination_x, destination_y)
+    return true if !destination_x.is_i? || !destination_y.is_i?
+    return true if destination_x < 1 || destination_x > 8 || destination_y < 1 || destination_y > 8
+    return true if !self.horizontal_or_vertical?(destination_x, destination_y)
+    return true if !self.diagonal?(destination_x, destination_y)
+    false
+  end
+
+  def horizontal_or_vertical_obstruction?(destination_x, destination_y)
+    return false unless self.horizontal_or_vertical?(destination_x, destination_y)
+    range_x = [destination_x, self.x_position].sort
+    range_y = [destination_y, self.y_position].sort
+    obstruction = Piece.where(y_position: ((range_y.first + 1)..(range_y.last - 1)), x_position: ((range_x.first + 1)..(range_x.last - 1))) # will always return something, even if it's an empty query
```

Comment:
Change `Piece.where` to `game.pieces.where`

That way, it will only look at pieces within the same game. Otherwise, it's ALL pieces in the database.